### PR TITLE
Sensor integration abstractions

### PR DESCRIPTION
**Pull request by Whathecode**
I introduced:

- `DeviceDataCollector` which is the abstraction for data collection on a _single_ device.
- `ConnectedDeviceDataCollector` extends on this and represents a _connected_ device. This is where I intend to later add battery and connectivity information (probably as a reactive flow).
- `DeviceDataCollectorFactory` which will be the entry point for developers to add integrations for sensors.
- `DeviceRegistrationStatus` can now be accessed through `StudyRuntimeStatus`.
- `ConnectedDeviceManager` which once a device has been registered (as exposed by `DeviceRegistrationStatus`) can be accessed from the `ClientManager` to provide access to device status, but does not expose subscription to data streams, etc.

Noteworthy is that there should only be one 'local' `DeviceDataCollector` which orchestrates data collection on the master device, and for each device with which there is a connection there should be an individual `ConnectedDeviceDataCollector`. There will still be issues on how to deal with aligning connections (BLE scanning, etc.), but we can focus on that in subsequent PRs.

While implementing this I discovered that the current requirement is verifying whether a connection can be made to a device _at a particular point in time_, and there was no need yet to have a continuous flow of connection information (including battery). After registering a device the client is expected to try to make a connection using that registration. But, device registration and device connection are separate processes and are thus exposed separately in the API (`DeviceRegistrationStatus` vs `ConnectedDeviceManager`).

What I still want to finish in this PR:

- [x] Inject `DeviceDataCollectorFactory` rather than `DataListener` into `ClientManager`. `DataListener` can be initialized privately from that. But, then `DataListener` will need to be passed when initializing `StudyRuntime` from a snapshot.
- [x] Write unit tests for `DataListener`.
- [x] Write proper 'success' unit tests (with study protocol with actual measures) for `StudyRuntime.tryDeployment`.
- [x] Add an initial 'battery' and 'connectivity' abstraction for `ConnectedDeviceDataCollector`. To access this we will have to see how to expose that in `ClientManager` and/or (?) `StudyRuntime`.
- [x] Can the `DeviceDataCollector` interfaces be made generic to strongly link the supported device registration to device descriptor the collector is designed for?

### CODE REVIEW

File: carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/data/DataListener.kt [hunk 1]
```diff
@@ -2,28 +2,45 @@ package dk.cachet.carp.client.domain.data
...
 import dk.cachet.carp.common.data.DataType
-import dk.cachet.carp.protocols.domain.devices.AnyDeviceDescriptor
 import dk.cachet.carp.protocols.domain.devices.DeviceRegistration
-import kotlin.reflect.KClass
+import dk.cachet.carp.protocols.domain.devices.DeviceType
 
 
 /**
- * Allows subscribing to [Data] of requested [DataType]s on a master device and connected devices.
+ * Allows subscribing to [Data] of requested [DataType]s on a master device and connected devices
```
Comment (patrickusiewicz) [inline carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/data/DataListener.kt:10]: I know this PRs focus is on internal/external sensors. But will the `DataListener` also be able to subscribe to data coming from survey answers? Either we should start adressing the future need for collecting survey data or create an issue.

Comment (Whathecode) [reply carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/data/DataListener.kt:10]: All data is treated equally. Will update documentation to state _"Allows subscribing to [Data] (e.g., sensors, surveys)"_ to clarify.

File: carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/ClientManager.kt [hunk 2]
```diff
@@ -27,11 +28,15 @@ abstract class ClientManager<
      */
     private val deploymentService: DeploymentService,
     /**
-     * Manages [Data] collection of requested [DataType]s for this master device and connected devices.
```
[the remark this answers is not in the document]
Comment (Whathecode) [reply carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/ClientManager.kt:30]: Aren't both javadoc, as in, both get picked up when generating documentation?  If so, this is purely a style preference.

As long as documentation is there and up-to-date, I can settle either way.

I think I mix the two. For one-line constructors/methods I document everything either in main comment using [] (which I think is the Kotlin recommendation), or use @param. For longer methods/constructors I use this notation.

I prefer documentation to be close to what is being documented (high cohesion, just like code). My hope is that when editing code people are more inclined to also update the documentation.

File: carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/ClientManager.kt [hunk 3]
```diff
@@ -85,7 +90,7 @@ abstract class ClientManager<
         // IllegalArgumentException's will be thrown here when deployment or role name does not exist, or device is already registered.
         val deviceRegistration = repository.getDeviceRegistration()!!
         val runtime = StudyRuntime.initialize(
-            deploymentService, dataCollector,
+            deploymentService, dataListener,
```
[the remark this answers is not in the document]
Comment (Whathecode) [reply carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/ClientManager.kt:93]: Idd, styling. :) I in fact had all on a separate line and thought it was _less_ meaningful than the current grouping. DI'd services are the first line, parameters are the second line.

**Review by patrickusiewicz — APPROVED** [review]

**Pull request merged**